### ISSUE TITLE
Support semantic scopes for non-`ref/call` edges

### ISSUE THREAD

**Issue by zrlk** [kythe/kythe#3946]
Generalize support for semantic scopes so we can get callgraph-like support for edges other than `ref/call`. This is most useful for `property/reads` and `property/writes` edges, as it will allow us to transitively blame changes to properties through the graph.

**Comment by zrlk** [comment]
#3945 adds property/reads and /writes edges.
#3707 adds support to the Documentation call.

**Comment by robinp** [comment]
Clarifying question: do I understand correctly that this lets edges other than ref/call participate in a callchain? Are there any other effects for API users / UIs? 

For example, I see a PR mentioning Documentation call support - but it seems more Decorations call? How would a UI use the scope returned as part of decorations?

**Comment by shahms** [comment]
It also potentially lets us match user expectations with language semantics for the existing `ref` edges, for example we've had more bugs filed than one would expect for calls "missing" from the call graph after refactoring Java from: `doFoo(p -> myfunc(p))` to `doFoo(this::myfunc)`, because the method reference isn't a call.

ISTR that it wasn't strictly feasible for all `ref` edges due to the volume of data, but it is one possibility.

**Comment by robinp** [comment]
@shahms: thanks for the context, makes sense. This was bothering me for a while - in "good old" languages, a call was a call. In Haskell though, very similar problems are to be faced as in your Java example.

I sense that one could make a sensible heuristic, but likely miss some cases.

1) Re infeasibility and data volume: what preprocessing does serving store for call chain info? Is it the backreferences together with a context function refs? Is it the latter extra refs that take up much volume?

The context is anyway emitted by the indexers, regardless of `ref` or `ref/call`. Is it that the table processing drops the context for plain `ref`s?

I'm asking since I always thought that serving a plain ref vs serving a callchain were not much difference, but apparently there's some tradeoff here that I miss.

2) As an alternative to precomputed callers: I assume the UI could issue kythe calls to reconstruct a callchain if demanded even for refs, isn't it?

**Comment by creachadair** [comment]
> In Haskell though, very similar problems are to be faced as in your Java example.

This is necessarily anecdotal, but I wouldn't expect most Haskell programmers to share that expectation. I think the reason we had complaints from some Java users was that they were "desugaring in their head", and wanted the new notation to follow the same path. I expect if you consult someone who started programming in Java after method references were released, they wouldn't share that expectation.

I think it's valuable to distinguish these two forms in the data model: Suppose the signature of `myfunc` changes. With `p -> myfunc(p)`, changes, you may have to edit that expression. However that is _never_ the case for `this::myfunc`. Indeed, that is one of the main points in having the special syntax for the method reference. While saving a couple bytes of typing is also nice, the real win is in saving the need to even consider that case. And this distinction falls naturally out of the model when call _syntax_ is used as the proxy for a call _site_. 

In Haskell you do have to decide what to do about partial applications, but that's orthogonal to the question about function references vs. calls: A partial application is still an application, it's only then a question of _what_ is being applied, and whether/how you can/should reference it.

**Comment by robinp** [comment]
@creachadair thanks, never thought about that benefit of the method reference form!

It seems the approach you take is semantic - what really is a call, what is a reference. I tend to think about debugging/UI, and ask the question, why would anyone _not_ want callchains for all the things?

For example, if say `foo` is a global constant, I would rather get (expandable) callchains for the use sites of foo (even if `foo` is cleanly not called). Because it is more practical to just expand the chain on the UI than going to the use sites one by one.

For Haskell (or anything with functions being first-class citizens, say JS too), there are tricky cases. One would expect `foo a b` and `flip foo b a` (which are equivalent) to behave similarly, but - if we introduce the distinction - the first case would be a ref/call with callchain, the second a plain ref.

Maybe the concept of a `callchain` doesn't fit languages with partial (or oversaturated) calls well, since the control flow is not as straightforward. But that would need more thinking from my part.

**Comment by creachadair** [comment]
> For example, if say `foo` is a global constant, I would rather get (expandable) callchains for the use sites of foo (even if `foo` is cleanly not called). Because it is more practical to just expand the chain on the UI than going to the use sites one by one.

That's a reasonable thing to want. Another constituency, however, does _not_ want to include non-calls among calls, because they are handled differently during (say) refactoring the definition of the function. That's why we have the distinction in the first place, in fact. I don't know that either viewpoint is obviously "better" (in the sense of being predominant). It may be something where we have to give the UI enough information in the model to choose which viewpoint to present.

> For Haskell (or anything with functions being first-class citizens, say JS too), there are tricky cases. One would expect `foo a b` and `flip foo b a` (which are equivalent) to behave similarly, but - if we introduce the distinction - the first case would be a ref/call with callchain, the second a plain ref.

We have to be somewhat careful about our view of "semantics", though. The semantics seen by the indexer is necessarily somewhat operational, not denotational: The compiler doesn't know the "meanings" of the phrases, only the rules for translating them for execution. As readers knowing the definition of `flip`, we can also predict that the _values_ of `foo a b` and `flip foo b a` are the same. The two are not "equivalent", however, in that they specify a different sequence of operational steps.

In this case the optimizer might partially evaluate those steps, and effectively rewrite the latter as the former. The type system can't guarantee that, however, so if we swap `flip` for some `wharrgarbl` we don't know the behaviour of, it matters that `wharrgarbl foo b a` and `foo a b` specify different steps. We know _a priori_ that `flip foo b a` will _expand_ to a call to `foo`, but we don't know that `wharrgarble foo b a` will—or perhaps it will yield several calls. All we can say for certain is that there is no call to `foo` _written_ at this location.

It is perhaps even _more_ important to the reader of the code that we not lose too much of the connection between what is written and what is captured by the index. The challenge is to let the user answer the questions they care about (e.g., "where are there actual calls to `foo` that I have to modify?" vs. "where are there locations where `foo` is mentioned, regardless of whether it's actually called?") without losing too much of that context.

All of which is mainly to argue in favour of the lexical view of calls, at least as far as building the model. Expanding that model to support different "filters" is IMO a UI consideration.

**Comment by robinp** [comment]
Thank you. This clarifies the expected semantics quite well.

The main open question for me then is operational: what special treatment does serving table construction give to `ref/call` that `call` doesn't get?

For example, would it be expensive to fetch the "calling context" for plain `ref`s when serving an xref request? Does it need special precomputation? Is it that the reverse `childof` lookup + details of that node are merged into the serving table beforehand? If so, am I right that the client could replicate that behavior, with the expense of the additional api calls, if wants to display callchain info for `ref`s as well?

**Comment by creachadair** [comment]
The main difference is that a `ref/call` has an extra `childof` edge pointing to its "caller", typically a `function`, that is blamed for the reference; other `ref` edges do not. Without the extra edge there's no good way to infer the enclosing context: The graph does not include syntax, except the minimal ranges spanned by anchors to locate identifiers and definition spans.

We did this because CodeSearch and other code browser UIs typically use the caller to give context to the call (seeing a callsite alone is typically not that informative without knowing where it is), and to support "upward" navigation in the call hierarchy. 

Of course, you _could_ support upward navigation for _any_ reference (and some newer ones might, though I've not seen it).  Historically, however, browser UIs only did it for function calls. Since variable references do not "expand" (e.g., you can't drill into a variable), they're typically displayed differently than calls.

And since the extra `@—childof—S` edges connecting call sites to their enclosing function are quite a bit of data, we only generate them for the cases where that navigation matters. Nothing in principle prevents assigning blame for _all_ references—it's mainly just a question of how much code structure you want to keep in the model and how much space it takes up.

In 2012 I did some experiments with capturing a more complete view of syntax along with the rest of the graph, and it wound up more than tripling the size of the index. That for Grok prior to the creation of Kythe, and used a compact shift-reduce sequence to encode each element—it'd be bigger if you indexed it for queries. At the time, that wasn't worth the cost, since no UI had any use for all all the nested lexical scopes, so we never bothered including it in the Kythe design. It could be done, but it'd need to justify its cost somehow. 

We never had any good user experiments, but without evidence I am skeptical that there is much value to a user browsing the lexical structure in the graph directly: The browser already has the code, and that is usually much easier to read than an abstracted tree view, once you get beyond coarse structures like top-level types and functions (the "outline view"). Encoding syntax could be much _more_ useful for programmatic inspection—but tools to do that are typically written with deep knowledge of the specific language they target, and don't benefit much from a language-agnostic view except as _output_ (e.g., surfacing results to a browser that doesn't know the IR of the analysis tool).

So, in summary: We made some design choices, and they could be revised. Nothing prevents us from revisiting all of these questions if there's cause (caveat that I do not have any real say in the final decision).

**Comment by robinp** [comment]
> Since variable references do not "expand" (e.g., you can't drill into a variable), they're typically displayed differently than calls.

I'm a bit confused. As you mention, call chains are used to navigate upward callers, so the ability to drill down seems orthogonal. Given a use site of a ref, or a call site of a call, one may want to see the context in which it is happening regardless.

Re tradeoffs: I see, thank you. With the new flag implemented in this issue, IIUC I could experiment with say instructing table construction to include context for `ref`s as well, and measure the output.

Maybe a full block syntax is indeed an overkill, but capturing the top-level function as a context could be useful. I will play around with it later. /cc @jinwoo

**Comment by creachadair** [comment]
> I'm a bit confused. As you mention, call chains are used to navigate upward callers, so the ability to drill down seems orthogonal. Given a use site of a ref, or a call site of a call, one may want to see the context in which it is happening regardless.

I only meant to point out that for calls you can go both up and down in the hierarchy, whereas for non-call references you can only go upward. The browsing tools we looked at, early on, either listed all references (including calls) as flat locations, or had a separate UI for call hierarchy specifically. Again, though, that's not set in stone.

Anecdotally, I expect users to care more often about the sequence of events when answering the question "how did I come to call this function?" as opposed to "how did I come to be setting this variable on this line?" But nothing at all prevents them being treated symmetrically AFAICS.

**Comment by jaysachs** [comment]
> I think the reason we had complaints from some Java users was that they were "desugaring in their head", and wanted the new notation to follow the same path. I expect if you consult someone who started programming in Java after method references were released, they wouldn't share that expectation.

I think what's happening here is that there are cases wehre programmers are interested in something, but strictly speaking it's not callgraphs (in the sense of "things identified by kythe as ref/call edges"). It's more like "where is this data actually used in interesting ways", "interesting" in terms of being other than simply passed verbatim to some other function ... and tracing "through" a method reference makes sense there.
